### ISSUE TITLE
ng serve behind HTTPS reverse proxy is failing to call https://localhost:4200

### ISSUE THREAD

**Issue by jandsu** [angular/angular-cli#2123]
Running `ng serve` behind an HTTPS reverse proxy is making requests on https://localhost:4200 when it should make them on http://localhost:4200
### OS?

Ubuntu 16.04
### Versions?

```
$ ng --version
angular-cli: 1.0.0-beta.11-webpack.9-4
node: 6.5.0
os: linux x64
```
### Repro steps
1. `ng new foo`
2. `cd foo`
3. `ng serve`
4. Then I load the page on https://mywebserver.mydomain.org which is the page served by a reverse proxy
   ==> observe the network tab in Chrome developer tools, it keeps making requests on

```
https://localhost:4200/sockjs-node/info?t=1473926195156
```

which fail when the same URL starting with http:// works

---

I have an HTTPS reverse proxy in front (to be able to route API calls - it needs to be HTTPS because I use the geolocation API). Here is an excerpt of the NGINX configuration file.

```
$ cat /etc/nginx/sites-enabled/my-test
server {
    listen 443 ssl;
        location / {
        proxy_pass http://localhost:4200;
    }
    ssl_certificate /etc/nginx/ssl/myserver.crt;
    ssl_certificate_key /etc/nginx/ssl/myserver-private.key;
}
```

**Comment by jandsu** [comment]
FWIW, same issue with beta 14
```
$ ng --version 
angular-cli: 1.0.0-beta.14
node: 6.5.0
os: linux x64

```

**Comment by filipesilva** [comment]
Calls like `https://localhost:4200/sockjs-node/info?t=1473926195156` are live reload pings, which I suppose just use the protocol the site is being served in (`https://` in your case). Those aren't really meant to be changed.

Not sure what I can do to help in your case, it's not really a problem with the CLI I believe. I'll flag it for community help.

Maybe documentation for the proxy feature helps? https://github.com/angular/angular-cli/pull/1896

**Comment by jandsu** [comment]
Please let me respectfully disagree.
I understand that those are pings on live reload... the correct behavior would be to ping on http://localhost:4200 regardless of the **scheme** and server I use (in my case it is https://mywebserver.mydomain.org).
The CLI is smart enough to replace the hostname with localhost and port with 4200, but it keeps the scheme... which is an error.

**Comment by LeviRosol** [comment]
I'll second @jandsu with another use case.

We utilize vagrant for out development environment and typically use a local `/etc/hosts` entry to map a name to a local IP that vagrant uses (ie: myapp.dev). Because of this we're using nginx to host the app from the vm (like @jandsu is doing, but for ssl).

We use vagrant so that developers who are working on multiple projects do not have to manage a projects dependancies on their machines. So that means we run things like `npm install` and `npm start` from the vm itself.

What this results in is the app using `localhost` for it's generated files. So when the reload makes it down to the developer machine, it's trying to use `localhost` for the reload domain vs `myapp.dev`

It does appear that supporting a custom host for reload was intended in serve based on what's showing here: https://github.com/angular/angular-cli/blob/3aa61d51c753b7467e33c358b1a7650e33b7c9c2/packages/angular-cli/commands/serve.ts#L56

I did try putting this to use, but it does not appear to have any impact on the host used in the reload script.

**Comment by LeviRosol** [comment]
via issue #1753 i found that by setting the host to 0.0.0.0 via `--host 0.0.0.0` my issue goes away completely. I'm now running the app on a vagrant box @ `localhost:4200`, serving it up via an nginx proxy @ `myapp.dev:1337` that I access from my base machine, and livereload seems to be making the request to `myapp.dev:1337` correctly. 

@jandsu I'm not sure if this will also resolve your issue since yours appears to be protocol vs host, but it's probably worth a shot.

Assuming my assumptions here are correct, it appears that the readme could use a little updating and / or the `live-reload-base-url` being fleshed out an utilized.

Hope this helps someone else!

**Comment by dhdanno** [comment]
I tried changing liveReloadPort to 443 with no apparent change on the client. I am reverse proxying https:// to 4200, so the desired behavior would be for the client to request a different port to what the server is hosting.

**Comment by mgol** [comment]
This is a Webpack issue, reproducible without Angular CLI. I created an issue, please vote for it (and add more info if you have some): https://github.com/webpack/webpack-dev-server/issues/763.

**Comment by xmeng1** [comment]
+1

**Comment by ccadieux** [comment]
We needed to add `ng serve --ssl` to get this to work.

We also had to add `--ssl-key` and `--ssl-cert` with the path to the key and certificate files.

**Comment by filipesilva** [comment]
Fixed by #3952.

**Comment by jandsu** [comment]
For anyone stumbling on this issue, there is indeed a `--public-host` option, however I personally don't wan't to commit the dev host name in the package.json command that runs ng serve since it is obviously different for each developer.

For now, I found this very dirty **hack**:
```
ng serve --public-host=fakehttp://0.0.0.0:0/
```
It works fine with Angular CLI 1.4.1. It only works because a webpack comitter wrote
`!!~self.location.protocol.indexOf("http")` rather than `self.location.protocol.indexOf("http")===0` (not sure why... but I hope noone changes that to make it less cryptic :) ).

The zero-port is required if you serve SSL on a non-standard port (which is my case, the 443 is used for the built rather than ng serve)
